processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferArrayGrouper.java

```diff
@@ -204,82 +223,57 @@ public void close()
   }
 
   @Override
-  public Iterator<Entry<KeyType>> iterator(boolean sorted)
+  public Iterator<Entry<Integer>> iterator(boolean sorted)
   {
-    return sorted ? sortedIterator() : plainIterator();
-  }
-
-  private Iterator<Entry<KeyType>> sortedIterator()
-  {
-    // Sorted iterator is currently not used because there is no way to get grouping key's cardinality when merging
-    // partial aggregation result in brokers and even data nodes (historicals and realtimes).
-    // However, it should be used in the future.
-    final BufferComparator comparator = keySerde.bufferComparator();
-    final List<Integer> wrappedOffsets = IntStream.range(0, cardinality + 1).boxed().collect(Collectors.toList());
-    wrappedOffsets.sort(
-        (lhs, rhs) -> comparator.compare(buffer, buffer, lhs + USED_FLAG_SIZE, rhs + USED_FLAG_SIZE)
-    );
-
-    return new ResultIterator(wrappedOffsets);
-  }
-
-  private Iterator<Entry<KeyType>> plainIterator()
-  {
-    return new ResultIterator(
-        IntStream.range(0, cardinality + 1).boxed().collect(Collectors.toList())
-    );
-  }
-
-  private class ResultIterator implements Iterator<Entry<KeyType>>
-  {
-    private static final int NOT_FOUND = -1;
-
-    private final Iterator<Integer> keyIndexIterator;
-    private int cur;
-    private boolean needFindNext;
-
-    ResultIterator(Collection<Integer> keyOffsets)
-    {
-      keyIndexIterator = keyOffsets.iterator();
-      cur = NOT_FOUND;
-      needFindNext = true;
+    if (sorted) {
+      throw new UnsupportedOperationException("sorted iterator is not supported yet");
     }
 
-    private int findNextKeyIndex()
+    return new Iterator<Entry<Integer>>()
     {
-      while (keyIndexIterator.hasNext()) {
-        final int index = keyIndexIterator.next();
-        if (isUsedKey(buffer, index * recordSize)) {
-          return index;
-        }
+      int cur = -1;
+      boolean findNext = false;
+
+      {
+        cur = findNext();
       }
-      return NOT_FOUND;
-    }
 
-    @Override
-    public boolean hasNext()
-    {
-      if (needFindNext) {
-        cur = findNextKeyIndex();
-        needFindNext = false;
+      @Override
+      public boolean hasNext()
+      {
+        if (findNext) {
+          cur = findNext();
+          findNext = false;
+        }
+        return cur != -1;
       }
-      return cur > NOT_FOUND;
-    }
 
-    @Override
-    public Entry<KeyType> next()
-    {
-      if (cur == NOT_FOUND) {
-        throw new NoSuchElementException();
+      private int findNext()
+      {
+        for (int i = cur + 1; i < cardinalityWithMissingValue; i++) {
+          if (isUsedSlot(i)) {
```

Comment:
I believe the usedFlag bitsets will be compact in most cases because BufferArrayGrouper is used for per-segment aggregation when the grouping key is an indexed string column. In this case the key is generated integer values by dictionary encoding, which should be compact.